sionna/channel/optical/edfa.py

```diff
@@ -0,0 +1,139 @@
+
+    For, e.g., a transparent link, the required gain per span is :math:`G = \exp\left(\alpha \Delta z\right)`.
+    The spontaneous emission factor calculates as :math:`n_\mathrm{sp}=\frac{F}{2}\frac{G}{G-1}`.
+    According to [A2012]_, [EKWFG2010]_ combined with [BGT2000]_, and [GD1991]_,
+    for the noise power spectral density of the EDFA per state of
+    polarization one obtains :math:`\rho_\mathrm{ASE}^{(1)} = n_\mathrm{sp}\left(G-1\right) h f_0=\frac{1}{2}G F h f_0`.
+    At simulation frequency :math:`f_\mathrm{sim}` the noise has a power of
+    :math:`P_\mathrm{ASE}^{(1)}=\sigma_\mathrm{n,ASE}^2=2\rho_\mathrm{ASE}^{(1)}\cdot f_\mathrm{sim}`
+    where the factor :math:`2` accounts for the unpolarized noise.
+    Here the :math:`()^{(1)}` means that this is the noise introduced by a
+    single EDFA.
+
+    This class inherits from the Keras `Layer` class and can be used as layer in
+    a Keras model.
+
+    Example
+    --------
+
+    Setting-up:
+
+    >>> edfa = EDFA(4.0, 2.0, 193.55e12, 1.0)
+
+    Running:
+
+    >>> # x is the optical input signal
+    >>> y = EDFA(([1.0+1.0j, 1.0+1.0j, 1.0+1.0j]))
+
+    Parameters
+    ----------
+        g : float
+            Amplifier gain
+
+        f : float
+            Noise figure
+
+        f_c : float
+            Carrier frequency :math:`f_\mathrm{c}` in :math:`(Hz)`
+
+        dt : float
+            Time step :math:`\Delta_t` in :math:`(s)`
+
+        dtype : Complex tf.DType
+            Defines the datatype for internal calculations and the output
+            dtype. Defaults to `tf.complex64`.
+
+    Input
+    -----
+        x : Tensor, tf.complex
+            Optical input signal
+
+    Output
+    -------
+        y : Tensor with same shape as ``x``, ``dtype``
+            Amplifier output
+    """
+    def __init__(self, g, f, f_c, dt, dtype=tf.complex64, **kwargs):
+        super().__init__(dtype=dtype, **kwargs)
+
+        self._complex_dtype = dtype
+        self._real_dtype = dtype.real_dtype
+
+        self._g = tf.cast(g, self._real_dtype)
+        self._f = tf.cast(f, self._real_dtype)
+        self._f_c = tf.cast(f_c, self._real_dtype)
+        self._dt = tf.cast(dt, self._real_dtype)
+
+        # Spontaneous emission factor
+        if self._g == 1.0:
+            self._n_sp = tf.cast(0.0, self._real_dtype)
+        else:
+            self._n_sp = self._f / tf.cast(
+                2.0, self._real_dtype) * self._g / (
+                                 self._g - tf.cast(1.0, self._real_dtype))
+
+        self._rho_n_ase = tf.cast(
+            self._n_sp * (self._g - tf.cast(1.0, self._real_dtype)) *
+            sionna.constants.H * self._f_c,
+            self._real_dtype)  # Noise density in (W/Hz)
+        self._p_n_ase = tf.cast(
+            2.0, self._real_dtype) * self._rho_n_ase * tf.cast(
+            1.0, self._real_dtype) / (self._dt)  # Noise power in (W)
+
+    def call(self, inputs):
```

Comment:
Did you ensure (e.g., via testing) that the EDFA/SSFM works with N-D inputs, i.e., only operates on axis=-1?

For example, let's say shape(x) = [bs, n], then x2 = tf.reshape(x, [bs/2, 2, n]) should generate the same output (after reshaping back to [bs, n]).

Reply:
The EDFA only multiplies the signal with a scalar that is broadcastable to all dimensions. Additionally, it adds noise with a given variance to all elements of the input tensor. As it has no memory it may operate on all axes (correct me if I am wrong).

For the SSFM there is a corresponding test (test_ssfm_batch_processing()) that checks if only axis=-1 is processed.